var prr = require('prr')

function init (type, message, cause) {
<<<<<<< HEAD
  if (!!message && typeof message != 'string') {
    message = message.message || message.name
  }
=======
>>>>>>> 3e7cc9eba236ff5c5205c34bae31be88b995768b
  prr(this, {
      type    : type
    , name    : type
      // can be passed just a 'cause'
    , cause   : typeof message != 'string' ? message : cause
<<<<<<< HEAD
    , message : message
=======
    , message : !!message && typeof message != 'string' ? message.message : message

>>>>>>> 3e7cc9eba236ff5c5205c34bae31be88b995768b
  }, 'ewr')
}

// generic prototype, not intended to be actually used - helpful for `instanceof`
function CustomError (message, cause) {
  Error.call(this)
  if (Error.captureStackTrace)
<<<<<<< HEAD
    Error.captureStackTrace(this, this.constructor)
=======
    Error.captureStackTrace(this, arguments.callee)
>>>>>>> 3e7cc9eba236ff5c5205c34bae31be88b995768b
  init.call(this, 'CustomError', message, cause)
}

CustomError.prototype = new Error()

function createError (errno, type, proto) {
  var err = function (message, cause) {
    init.call(this, type, message, cause)
    //TODO: the specificity here is stupid, errno should be available everywhere
    if (type == 'FilesystemError') {
      this.code    = this.cause.code
      this.path    = this.cause.path
      this.errno   = this.cause.errno
      this.message =
        (errno.errno[this.cause.errno]
          ? errno.errno[this.cause.errno].description
          : this.cause.message)
        + (this.cause.path ? ' [' + this.cause.path + ']' : '')
    }
    Error.call(this)
    if (Error.captureStackTrace)
<<<<<<< HEAD
      Error.captureStackTrace(this, err)
=======
      Error.captureStackTrace(this, arguments.callee)
>>>>>>> 3e7cc9eba236ff5c5205c34bae31be88b995768b
  }
  err.prototype = !!proto ? new proto() : new CustomError()
  return err
}

module.exports = function (errno) {
  var ce = function (type, proto) {
    return createError(errno, type, proto)
  }
  return {
      CustomError     : CustomError
    , FilesystemError : ce('FilesystemError')
    , createError     : ce
  }
}
